text/0000-offline-buildpackages.md

```diff
@@ -0,0 +1,463 @@
+
+To do so we introduce the asset cache, a distribution artifact that provides vendored assets for buildpacks, and platform mechanisms to provide vendored assets to buildpacks during `build` and `detect`. We also add a top level `[[assets]]` array in `buildpack.toml`
+
+# Motivation
+[motivation]: #motivation
+
+- Simplify artifacts needed to achieve offline builds.
+- Enable registry deduplication of buildpack assets
+- Separate buildpacks from vendored assets.
+
+# What it is
+[what-it-is]: #what-it-is
+
+### Define the target persona:
+Buildpack user, Platform operator, Buildpack author
+
+### Explaining the feature largely in terms of examples.
+
+### Asset Cache Definition
+We define the asset cache artifact. It is a reproducible OCI image. Each layer in this image contains one or more `asset` files a buildpack may contribute during a future build. Each of these `asset` files will be made available at `/cnb/assets/<asset-sha256>` in the build container.
+
+##### General Layer Structure:
+```
+<layer1> ┳━ /cnb/assets/<java11-asset-sha256>
+         ┗━ /cnb/assets/<java13-asset-sha256>
+
+...
+<layern> ━━ /cnb/assets/<java15-asset-sha256>
+```
+
+### Asset Cache Image Labels
+Asset caches will have two labels. An`io.buildpacks.asset.cache.metadata` label that contains a human readable name.
+``` json
+{
+  "name": "asset-org/asset-name"
+}
+```
+
+Asset caches will also have a `io.buildpacks.asset.cache.layers` Label. This label maps individual `assets` to the containing layer using `layerDiffID`
+
+##### General Format:
+```json
+{
+  "<asset-sha256>": {
+    "name": "(optional)",
+    "id": "(required)",
+    "version": "(required)",
+    "layerDiffID": "(required)",
+    "uri": "(optional)",
+    "stacks": [
+      "(optional)",
+      "..."
+    ],
+    "metadata": {
+      "(optional)": "(optional)"
+    }
+  }
+}
+```
+
+##### Example:
+
+``` json
+{
+  "<java11-asset-sha256>": {
+    "id": "java",
+    "version": "11.0.9",
+    "name": "java11",
+    "layerDiffId": "<layer1-diffID>",
+    "uri": "https://path/to/java11.tgz",
+    "stacks": ["io.buildpacks.stacks.bionic"],
+    "metadata": {}
+  },
+  "<java13-asset-sha256>": {
+    "id": "java",
+    "version": "13.0.2",
+    "layerDiffId": "<layer1-diffID>",
+    "stacks": ["io.buildpacks.stacks.bionic"],
+    "metadata": {
+      "license": "..."
+    }
+  },
+  "<java15-asset-sha256>": {
+    "id": "java",
+    "version": "15.0.2",
+    "stacks": ["io.buildpacks.stacks.bionic"],
+    "layerDiffId": "<layerk-diffID>",
+  }
+}
+```
+
+### Buildpack.toml changes
+
+In order to maintain a link between buildpacks and the assets they may provide we add a top level `[[assets]]` array to the `buildpack.toml` file.
+
+##### General Format
+
+```toml
+[buildpack]
+  id = "paketo-buildpacks/node-engine"
+  version = "0.0.1"
+ 
+[[assets]]
+  sha256 = "(required)"
+  name = "(optional)"
+  id   = "(required)"
+  version = "(required)"
+  uri = "(optional)"
```

Comment:
What are the cases where a `uri` would not be provided. An asset that is not available for public download? In that case is an asset override required to create the asset cache? What happens if I try to create a builder from this buildpack without an asset cache? Does it fail to create the builder? Or will we wait for the buildpack to fail in cases where the asset is needed?

Reply:
I believe VMware has a internal dynatrace buildpack that constructs the `uri` dynamically. Thought it might be useful to still have an entry in `buildpack.toml` to indicate that an asset is downloaded as part of buildpack execution, or use it for some `sha256` validation if needed.